### ISSUE TITLE
Kernel.system calls cause server to stop working

### ISSUE THREAD

**Issue by anex-at-work** [grpc/grpc#38210]
### What version of gRPC and what language are you using?

gRPC >= 1.65.0, Ruby >= 3.1.0

### What operating system (Linux, Windows,...) and version?

Various versions of Linux, including containers based on (for example), `ruby:3.X.Y-slim-bullseye`, MacOS (latest).  

### What runtime / compiler are you using (e.g. python version or version of GCC)

MRI

### What did you do?

Unit test: https://github.com/anex-at-work/grpc/commit/e8d7dc32095aa82fe2223c6f0e77a7787e5192ed (line [180](https://github.com/anex-at-work/grpc/commit/e8d7dc32095aa82fe2223c6f0e77a7787e5192ed#diff-d2b230f982304c473c5174c59efa17459047f8359edcf16c792de1e7fc3172cbR180)). 

Literally, any implementation of server handler that is using either `Kernel.system` or [method-i-60](https://ruby-doc.org/core-3.1.1/Kernel.html#method-i-60) methods (e.g.:
```
`echo "test"`
```
)  causes the next issue in unit tests:

```
GRPC::Cancelled:
       1:CANCELLED. debug_error_string:{UNKNOWN:Error received from peer  {created_time:"2024-12-02T00:07:53.366634941+00:00", grpc_status:1, grpc_message:"CANCELLED"}}
     # ./src/ruby/lib/grpc/generic/active_call.rb:29:in `check_status'
     # ./src/ruby/lib/grpc/generic/active_call.rb:189:in `attach_status_results_and_complete_call'
     # ./src/ruby/lib/grpc/generic/active_call.rb:383:in `request_response'
     # ./src/ruby/lib/grpc/generic/client_stub.rb:180:in `block in request_response'
     # ./src/ruby/lib/grpc/generic/interceptors.rb:170:in `intercept!'
     # ./src/ruby/lib/grpc/generic/client_stub.rb:179:in `request_response'
     # ./src/ruby/lib/grpc/generic/service.rb:171:in `block (3 levels) in rpc_stub_class'
     # ./src/ruby/spec/generic/rpc_server_spec.rb:599:in `block (5 levels) in <top (required)>'
     # ./src/ruby/spec/generic/rpc_server_spec.rb:599:in `block (4 levels) in <top (required)>'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example.rb:263:in `instance_exec'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example.rb:263:in `block in run'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example.rb:511:in `block in with_around_and_singleton_context_hooks'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example.rb:468:in `block in with_around_example_hooks'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/hooks.rb:486:in `block in run'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/hooks.rb:624:in `run_around_example_hooks_for'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/hooks.rb:486:in `run'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example.rb:468:in `with_around_example_hooks'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example.rb:511:in `with_around_and_singleton_context_hooks'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example.rb:259:in `run'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example_group.rb:646:in `block in run_examples'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example_group.rb:642:in `map'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example_group.rb:642:in `run_examples'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example_group.rb:607:in `run'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example_group.rb:608:in `block in run'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example_group.rb:608:in `map'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example_group.rb:608:in `run'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example_group.rb:608:in `block in run'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example_group.rb:608:in `map'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/example_group.rb:608:in `run'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/runner.rb:121:in `block (3 levels) in run_specs'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/runner.rb:121:in `map'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/runner.rb:121:in `block (2 levels) in run_specs'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/configuration.rb:2092:in `with_suite_hooks'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/runner.rb:116:in `block in run_specs'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/reporter.rb:74:in `report'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/runner.rb:115:in `run_specs'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/runner.rb:89:in `run'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/runner.rb:71:in `run'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/lib/rspec/core/runner.rb:45:in `invoke'
     # /usr/local/rvm/gems/default/gems/rspec-core-3.13.1/exe/rspec:4:in `<main>'
```

If I'm right this behaviour caused by https://github.com/grpc/grpc/commit/d4b5e12f8ecf22b506b2831a4614c62a83458a31 (src/ruby/ext/grpc/rb_completion_queue.c specifically: rolling back those changes get expected/normal behavior back).

### What did you expect to see?

No cancelation in accepting requests, and any system call should not interfere with the server's work.

### What did you see instead?

In the gRPC server any call of `Kernel.system` causes an infinite loop of the next messages:
```
{"time":TIMESTAMP,"level":"WARN","type":"server call failed: request_call completion failed"}
```

The server stops accepting any requests and requires a restart.

### Anything else we should know about your project / environment?

It's crucial sometimes to run something from a shell that is impossible (during regulation, for example, reasons) to reimplement inside the code, and before version `1.65.0` everything works just fine.

**Comment by Montana** [comment]
You can isolate the system call by running it in a separate thread, preventing it from blocking the main event loop of the gRPC server. This allows the server to continue handling requests without being blocked by the Kernel.system call. Here's an example:

```ruby
def run_system_command(command)
  Thread.new do
    Kernel.system(command)
  end
end

class EchoService < EchoService::Service
  def Echo(request, call)
    run_system_command("echo 'test'")  # Use thread to prevent blocking
    call.send_response(EchoResponse.new(message: "Test complete"))
  end
end
```
By doing this, the system command will be executed in the background, and the main thread (which is handling the gRPC calls) will not be blocked.

Cheers,
Montana

**Comment by anex-at-work** [comment]
Thank you, @Montana! I appreciate your help. Your suggested method would work if we didn’t need to capture the result of the system/shell call. Unfortunately, as long as the thread needs to remain synchronized in any form, the issue inevitably reappears. (Just to clarify, this behavior occurs only in gRPC versions 1.65 and above.)

Please correct me if I’m wrong, but the test results suggest that threads won’t resolve the issue. 😔

Thanks again!

**Comment by apolcyn** [comment]
I just gave this a try with 1.69, tweaking greeter_server.rb in our examples to run `echo RPC` in the rpc handler.

Worked as expected and did not repro.

I have trouble seeing what this could be caused by or how [referenced commit](https://github.com/grpc/grpc/commit/d4b5e12f8ecf22b506b2831a4614c62a83458a31) relates.

Please provide a simple repro.

**Comment by anex-at-work** [comment]
@apolcyn Thank you for your response!
You can easily reproduce the issue by using a server definition like this:
```ruby
# GreeterServer is simple server that implements the Helloworld Greeter server.
class GreeterServer < Helloworld::Greeter::Service
  # say_hello implements the SayHello rpc method.
  def say_hello(hello_req, _unused_call)
    Kernel.system('echo "test"')
    Helloworld::HelloReply.new(message: "Hello #{hello_req.name}")
  end
end
```
Please note that to run this, I had to recompile the proto files. Without recompiling, [this file](https://github.com/grpc/grpc/blob/master/examples/ruby/lib/helloworld_pb.rb) causes 
```
undefined method `build' for an instance of Google::Protobuf::DescriptorPool` for the version of gRPC >= `1.65.0`).
```

With the above server definition, the server responds to the first call like this: (for a client)
```
"Greeting: Hello world"
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1736639275.158268   22732 chttp2_transport.cc:1204] ipv6:%5B::1%5D:50051: Got goaway [2] err=UNAVAILABLE:GOAWAY received; Error code: 2; Debug Text: Cancelling all calls {created_time:"2025-01-11T23:47:55.158266722+00:00", http2_error:2, grpc_status:14
```
For version `1.69.0`, the first call does not show any issues on the server side. However, the second call fails, and the client responds with:
```
bundle exec ./greeter_client.rb 
ERROR: 14:failed to connect to all addresses; last error: UNKNOWN: ipv4:127.0.0.1:50051: Failed to connect to remote host: Connection refused. debug_error_string:{UNKNOWN:Error received from peer  {created_time:"2025-01-11T23:50:39.259592739+00:00", grpc_status:14, grpc_message:"failed to connect to all addresses; last error: UNKNOWN: ipv4:127.0.0.1:50051: Failed to connect to remote host: Connection refused"}}
```
![image](https://github.com/user-attachments/assets/f92f7449-f9ed-4325-b714-e91c97248422)

If needed, I can provide a repository, but it is essentially a mirror of the [current example](https://github.com/grpc/grpc/tree/master/examples/ruby) folder.

**Comment by chadlwilson** [comment]
Did you have a chance to repro this @apolcyn ? Can we help further?

I've been stuck on grpc `1.64` due to an unknown problem for quite a long time now, before today finally tracking it down to this - thanks for raising!

FWIW, this only seems to fail in this way for me on Ruby `3.3` and `3.4`, across both Linux and MacOS - although @anex-at-work seems to think it's `Ruby >= 3.1.0`. Are you sure you've seen it on Ruby 3.1 and 3.2?

**Comment by apolcyn** [comment]
aha, yes I can repro *with* ruby-3.4  (FWIW I could not repro with ruby 3.2)

With a cursory look https://github.com/grpc/grpc/commit/d4b5e12f8ecf22b506b2831a4614c62a83458a31 definitely looks related.

That commit changed our `rb_thread_call_without_gvl` "unblocking func" to itself initiate server shutdown.

Likely something changed in a ruby 3.X version that triggers `rb_thread_call_without_gvl` unblocking func to run when we shell out (e.g. perhaps b/c of a signal for the child process).

We were probably "getting away with this" before that grpc patch, by taking [this code path](https://github.com/grpc/grpc/blob/b64756acca2eb942c97a416850ce5ab95a544d3e/src/ruby/lib/grpc/generic/rpc_server.rb#L474) to avoid server shutdown.

I'll try to take a deeper look later this week.

**Comment by chadlwilson** [comment]
Could it perhaps relate to https://github.com/ruby/ruby/commit/7fd53eeb46db261bbc20025cdab70096245a5cbe ?

**Comment by Montana** [comment]
Hi everyone, 

To diagnose whether the changes in `process.c(commit d4b5e12)` are causing the issue, try rolling them back and testing if this resolves the unexpected behavior. If reverting these modifications restores normal execution without canceled gRPC requests, it confirms that the issue originates from these changes. 

Additionally, replacing `Kernel.system` with `Process.spawn` provides better control over subprocess execution, reducing the risk of issues related to `SIGCHLD` handling. 

Unlike `Kernel.system`, Process.spawn allows for more flexible process management, preventing unintended interference with the gRPC server. Furthermore, explicitly calling `Process.waitpid` within your implementation ensures that child processes are properly reaped, avoiding the accumulation of zombie processes and reducing the chances of infinite loops or server instability. 

If Kernel.system or exec is causing issues, manually handling waitpid is a necessary step to maintain stable process execution in your gRPC environment.

I wrote a patch that might work: 

```c
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>

void reap_child_processes(int signo) {
    int saved_errno = errno;
    while (waitpid(-1, NULL, WNOHANG) > 0);
    errno = saved_errno;
}

void restore_waitpid_behavior() {
    struct sigaction sa;
    sa.sa_handler = reap_child_processes;
    sigemptyset(&sa.sa_mask);
    sa.sa_flags = SA_RESTART | SA_NOCLDSTOP;

    if (sigaction(SIGCHLD, &sa, NULL) < 0) {
        perror("sigaction failed");
        exit(1);
    }
}

void run_command(const char *cmd) {
    pid_t pid = fork();
    if (pid < 0) {
        perror("fork failed");
        exit(1);
    } else if (pid == 0) {
        execl("/bin/sh", "sh", "-c", cmd, (char *)NULL);
        perror("exec failed");
        exit(1);
    }
}

void initialize_fix() {
    restore_waitpid_behavior();
}
```

With this fix, `Kernel.system` calls will no longer interfere with gRPC request handling, ensuring that subprocess execution does not disrupt server operations. The gRPC server will continue to function smoothly without hanging or entering an infinite loop, maintaining stability even when executing external commands. Additionally, the issue of `CANCELLED` errors caused by improper subprocess management will be resolved, preventing unexpected failures and allowing the server to handle requests reliably.

_Michael Mendy_

**Comment by anex-at-work** [comment]
@chadlwilson, @apolcyn My sincere apologies for the confusion regarding Ruby versions. At this point, I don’t recall the source of the 3.1 reference. I have checked again and can confirm that I am unable to reproduce the issue with Ruby 3.1 or 3.2. I have updated the description accordingly.

@Montana Thank you for the patch! Unfortunately, we are unable to remove or replace that shell call, modify official dependencies, or introduce additional libraries - especially since external requirements prevent us from maintaining our own gem (even if it may seem impractical or questionable).

**Comment by chadlwilson** [comment]
Thanks @anex-at-work that helps validate we are seeing the same root cause.

I also haven't been able to figure out an appropriate workaround.

**Comment by chadlwilson** [comment]
Any luck with this one @apolcyn ? 😬

**Comment by apolcyn** [comment]
Sorry have not gotten a chance yet to give this proper attention.

cc @HannahShiSFB

**Comment by chadlwilson** [comment]
@apolcyn is there anything we can do to help move this along, or possible workarounds to experiment with?

**Comment by apolcyn** [comment]
Sorry for the delay here. Still haven't been able to give attention to this.

Quick summary of where this is at though:

Note, ideally we could just revert https://github.com/grpc/grpc/pull/36916, but that itself is a fix forward for another issue that can cause a thread to hang indefinitely upon process shutdown).

But, now that Event Engine has been enabled (ref https://github.com/grpc/grpc/blob/c68886d1c5e9a4014523ead76bc1224ed94560d7/src/core/lib/experiments/rollouts.yaml#L57 and other `event_engine_` experiments), we can probably go ahead and remove the aforementioned [thread](https://github.com/grpc/grpc/blob/c68886d1c5e9a4014523ead76bc1224ed94560d7/src/ruby/ext/grpc/rb_channel.c#L780) entirely (we need to confirm the EE experiment is rollback safe, though).

After that thread is removed, then we could do a small revert to go back to the `completion_queue_pluck` behavior that we had before https://github.com/grpc/grpc/pull/36916 (i.e. go back to the timer-based polling loop we had earlier).

Then, grpc would be fixed for these new ruby versions.

----------------

Separately, the newer ruby behavior with spurious invocations of the unblocking func for [rb_thread_call_without_gvl](https://github.com/ruby/ruby/blob/4646ab89179aa3dac917a16aaa16533548ad33d9/thread.c#L1592C4-L1592C30) still *feels* like an issue (unless I'm missing something). If that could end up being fixed, then we could keep the behavior from https://github.com/grpc/grpc/pull/36916 around.

**Comment by apolcyn** [comment]
Spent some time on this, with WIP in https://github.com/grpc/grpc/pull/39409 FYI

But pausing that on a couple things (confirming EE is fully enabled, and fixing issues found while re-running forking tests)

**Comment by alhafoudh** [comment]
I found this issue while I was trying to solve a similar problem where I wanted to start `binding.irb` session during GRPC call. Can it be related?

**Comment by apolcyn** [comment]
https://github.com/grpc/grpc/pull/39409 merged and should fix this (in the 1.74 release)

**Comment by chadlwilson** [comment]
Fantastic! Let me have a go with this on a dev build if possible in my env.

**Comment by chadlwilson** [comment]
Unfortunately this is quite the challenge to get working through our automated functional tests as it apparently uses all 14GB of GitHub Actions runner disk space trying to build the grpc gem on a normal bundle install (deep clones of all the large repos probably don't help).

Might need some bundler or build incantations to get it to use less disk space during extension compile.

**Comment by apolcyn** [comment]
@chadlwilson not sure if it helps - there are nightly builds off the master branch, available under https://packages.grpc.io/

**Comment by chadlwilson** [comment]
Thanks @apolcyn - managed to get things to work off source builds in the end. So far it looks like `1.74.0.dev` @ https://github.com/grpc/grpc/commit/11d21af06249796909bd9be51e9b0e09668710a6 fixes this for us across Ruby 3.3 -> 3.4 and also is working fine on 3.1 -> 3.2. The functional tests that were failing after `Kernel.system` calls before are working now.

**Comment by apolcyn** [comment]
That's great to here :) 

Thanks for confirming @chadlwilson

**Comment by chadlwilson** [comment]
I think this ticket, #39147 and possibly #38442 can (optimistically?) be closed now?

**Comment by apolcyn** [comment]
Yep, closing as the fix will be released in 1.74

**Comment by MrShemek** [comment]
@apolcyn when can we expect 1.74 to be released?
